### ISSUE TITLE
Bug: Weird rerender behavior with render prop and context

### ISSUE THREAD

**Issue by sidecus** [facebook/react#19208]
I am seeing some weird re-render behavior when using render prop and react context.
Here is my sample setup to show some context based tooltip title on a button:
1. Function component with useContext directly
2. A component referencing another function component which uses useContext and renders with render props (eventual goal is to provide some hooks based functionality to class components)
3. A memoized version of # 2 setup with React.memo
4. Instead of rendering with JSX syntax, set the tooltip to a function result similar as # 2 above

I have 4 different buttons using above setup respectively. When I hover each button, here is the behavior I am observing:
1. The function component always rerenders - as expected
2. The wrapper component and nested function component always rerenders - as expected as well
3. Memoization doesn't take effect, wrapper and nested components always rerender - **unexpected**
4. The wrapper component doesn't rerender, but the nested component always rerenders even though context doesnt' change - **unexpected** - **EDIT**: this is not a real issue but due to I misunderstood the component hierarchy in this case .

Runnable code sample is as below.

React version: 16.9.0

## Steps To Reproduce
See above. # 3  and # 4 are unexpected and it'll be great if someone can help shed some light on what's going on.

## Link to code example:
[Sample code](https://github.com/sidecus/reactstudy/blob/master/src/RenderPropHooks/RenderPropHooks.tsx)

You can also run the repo directly - once started click on tab "RenderProp and ContextHook".
Here is a [github page deployment](https://sidecus.github.io/reactstudy/).

**Comment by vkurchatkin** [comment]
I don't see anything unexpected here:

Here `React.memo` is used incorrectly, you are comparing prop objects instead of comparing props one by one. You should just omit the second argument.

```js
const MemoedTitleWithRenderProp = React.memo(TitleWithRenderProp, (prev, next) => {
    return prev === next;
});
```


This component rerenders because because its parent component is rerendered. You need to use React.memo to avoid that.

```js
const AddNumberToContext = ({toAdd, children} : {toAdd: number, children: (result: number) => JSX.Element}) => {
    const context = React.useContext(Context);
    console.log('AddNumberToContext rendering');
    return children(context + toAdd);
};
```

**Comment by sidecus** [comment]
Thank you @vkurchatkin for taking the time looking into this.

For the former sorry I committed the wrong code by mistake when playing around different possibilities. I actually tried both:
```
const MemoedTitleWithRenderProp = React.memo(TitleWithRenderProp);
```
and
```
const MemoedTitleWithRenderProp = React.memo(TitleWithRenderProp, (prev, next) => {
    console.log("propsAreEqual is called");
    return prev.toAdd === next.toAdd;
});```
It rerenders regardless when parent rerenders and the propsAreEqual call back is not even called.

For the latter, you are right, now thinking a bit more I realized that it has one component less in the hierarchy and AddNumberToContext rerenders because parent rerenders as you mentioned, not because of context update.
